dev/doc/parsing.md

```diff
@@ -0,0 +1,378 @@
+very specific to Coq (not so similar to Camlp5):
+
+* OCaml code - OCaml code enclosed in curly braces, which is copied verbatim to the generated `*.ml` file
+
+* Comments - comments in the `*.mlg` file in the form `(* ... *)`, which are dropped.  Comments
+  in OCaml code are preserved.
+
+* `DECLARE_PLUGIN "*_plugin"` - associates the file with a specific plugin, for example "ltac_plugin"
+
+* `GRAMMAR EXTEND` - adds additional nonterminals and productions to the grammar and declares global
+  nonterminals referenced in the `GRAMMAR EXTEND`:
+
+  ```
+  GRAMMAR EXTEND Gram
+    GLOBAL:
+      bignat bigint ...;
+    <nonterminal definitions>
+  END
+  ```
+
+  TODO: What is `Gram` above?  Worth explaining?
+
+  Global nonterminals are declared in `pcoq.ml`, e.g. `let bignat = Entry.create "Prim.bignat"`.
+  All the `*.mlg` files include `open Pcoq` and often its modules, e.g. `open Pcoq.Prim`.
+
+* `VERNAC COMMAND EXTEND` - adds new command syntax by adding productions to the
+  `command` nonterminal.  For example:
+
+  ```
+  VERNAC COMMAND EXTEND ExtractionLibrary CLASSIFIED AS QUERY
+  | [ "Extraction" "Library" ident(m) ]
+    -> { extraction_library false m }
+  END
+  ```
+
+  TODO: explain "ExtractionLibrary", CLASSIFIED AS, CLASSIFIED BY, "{ tactic_mode }", STATE
+
+  Productions here are represented with alternate syntax, described later.
+
+  It's also possible to add productions to `command` with `GRAMMAR EXTEND` but
+  without the additional fields.
+
+  TODO: Which is the best practice?
+
+* `VERNAC { ... } EXTEND` - TODO
+
+* `TACTIC EXTEND` - adds new tactic syntax by adding productions to `simple_tactic`.
+  For example:
+
+  ```
+  TACTIC EXTEND btauto
+  | [ "btauto" ] -> { Refl_btauto.Btauto.tac }
+  END
+  ```
+
+  TODO: "btauto", DEPRECATED, LEVEL
+
+* `ARGUMENT EXTEND` - defines a new nonterminal
+
+  ```
+  ARGUMENT EXTEND ast_closure_term
+       PRINTED BY { pp_ast_closure_term }
+       INTERPRETED BY { interp_ast_closure_term }
+       GLOBALIZED BY { glob_ast_closure_term }
+       SUBSTITUTED BY { subst_ast_closure_term }
+       RAW_PRINTED BY { pp_ast_closure_term }
+       GLOB_PRINTED BY { pp_ast_closure_term }
+    | [ term_annotation(a) constr(c) ] -> { mk_ast_closure_term a c }
+  END
+  ```
+
+  TODO: what is the purpose of this?
+  TODO: what are the various keywords for?  And TYPED AS?
+
+  Creates a "wit_" symbol
+
+* `VERNAC ARGUMENT EXTEND` - (part of `argument_extend` in the yacc grammar) defines
+  productions for a single nonterminal.  For example:
+
+  ```
+  VERNAC ARGUMENT EXTEND language
+  PRINTED BY { pr_language }
+  | [ "Ocaml" ] -> { let _ = warn_deprecated_ocaml_spelling () in Ocaml }
+  | [ "OCaml" ] -> { Ocaml }
+  | [ "Haskell" ] -> { Haskell }
+  | [ "Scheme" ] -> { Scheme }
+  | [ "JSON" ] -> { JSON }
+  END
+  ```
+
+  TODO: creates wit_ variable--makes it possible to refer to when using alternative prodn syntax?
```

Comment:
wit variables are used by `VERNAC / TACTIC EXTEND` that uses those arguments.

Reply:
IIUC, it seems tedious to add _every_ new nonterminal with `ARGUMENT EXTEND` instead of, say, using `GRAMMAR EXTEND`.  Is `ARGUMENT EXTEND` always preferable to `GRAMMAR EXTEND`?  If not, when should it be used?  Seems to me it might be a struggle to get everyone to use it, and considerable work to (perhaps) convert existing code to use it as well.

Also, FWIW, "ARGUMENT" doesn't seem like a very descriptive name for what it does.

Reply:
It is *critical* to use those macros. `GRAMMAR EXTEND` is just about syntax, when `ARGUMENT EXTEND` is about semantics.

Reply:
OK.  How can I briefly explain which nonterminals it's critical for, because there are tons of nonterminals defined in `GRAMMAR EXTEND`?  If people understand the reason for the guideline they're more likely to follow it.

Reply:
Except if you reimplement a tactic language, you shouldn't use `GRAMMAR EXTEND`.